master/internal/trial.go

```diff
@@ -741,11 +741,7 @@ func (t *trial) processContainerTerminated(
 		})
 	}
 
-	status := agent.ContainerError(agent.AgentError, errors.New("no error status provided"))
-	if s := msg.ContainerStopped; s.Failure != nil {
-		status = *s
-	}
-	if c == nil || c.IsLeader() || status.Failure != nil {
+	if c == nil || c.IsLeader() || msg.ContainerStopped.Failure != nil {
```

Comment:
Let's put a comment that explains in plain english what we are trying to do with each clause of this if statement.

This whole bug would probably have been avoided if we just explained in a comment what we were trying to do in the first place.

Reply:
Added a comment in here describing the three cases.

Reply:
nit: we are actually terminating the task if the leader exited at all, not just if it was canceled

Also, given the complexity of the state machine, it might be nice to explain under what conditions "the container never started".

Reply:
Done.

Reply:
looks good

Reply:
Perhaps the "if the container never started" should also get a tiny bit more explanation, since it's not necessarily immediately clear why the task should be terminated in that case (and I say that as the one who added that condition). I'm thinking "Terminate the task if this container never actually began executing (since the gang will never be able to completely start), if the leader of the gang exited, or if any of the containers exited with a failure."

Reply:
Done.